micrometer-observation/src/test/java/io/micrometer/observation/ObservationTests.java

```diff
@@ -150,6 +153,62 @@ void scopedCheckedSpecificException() {
         }
     }
 
+    @Test
+    void observe() {
+        String result;
+        // with supplier
+        result = Observation.start("service", registry).observe(() -> "foo");
+        assertThat(result).isEqualTo("foo");
+
+        // with function
+        result = Observation.start("service", registry).observe((context) -> "bar");
+        assertThat(result).isEqualTo("bar");
+
+        // with runnable
+        AtomicBoolean called = new AtomicBoolean();
+        Observation.start("service", registry).observe(() -> {
+            called.set(true);
+        });
+        assertThat(called).isTrue();
+    }
+
+    @Test
+    void observeWithFunction() {
+        CustomContext context = new CustomContext();
+        AtomicReference<CustomContext> contextHolder = new AtomicReference<>();
+        String result = new SimpleObservation("service", registry, context).observe((ctx) -> {
+            CustomContext customContext = (CustomContext) ctx;
+            customContext.hello();
+            contextHolder.set(customContext);
+            return "foo";
+        });
```

Comment:
I'm not sure if it is very important but I would try to do this without initializing the `Context`. I guess that was Artem's main issue: he does not want to create a context if the observation is a noop.

```suggestion
        String result = new SimpleObservation("service", registry, CustomContext::new).observe((ctx) -> {
            CustomContext customContext = (CustomContext) ctx;
            customContext.setMessage("test");
            return "foo";
        });
```
And then verify if `customContext.getMessage()` equals to `"test"`.
For this, `CustomContext` needs a `setMessage`/`getMessage` on a `static` field (I'm not sure how ugly is that). Or you can get a reference to the observation, call `getContext` on it and do the verification against that.